install/installer/pkg/common/common.go

```diff
@@ -350,11 +350,26 @@ func RepoName(repo, name string) string {
 	if err != nil {
 		panic(fmt.Sprintf("cannot parse image repo %s: %v", ref, err))
 	}
-	return pref.String()
+
+	prefString := pref.String()
+
+	if cfg.Repository != GitpodContainerRegistry {
+		// If not in the Gitpod registry, don't use namespaces unless specified in the config.
+		// This is to match the KOTS registry format of no namespace in the name
+		s := strings.Split(prefString, "/")
+
+		noNamespace := []string{
+			cfg.Repository,
+			s[len(s)-1],
+		}
+		return strings.Join(noNamespace, "/")
+	}
```

Comment:
Could you give some background on what's happening here and what problem it solves?
It strikes me that this
1. is the cause of the `cfg` inclusion/need for passing `cfg.Repository` along,
2. possibly unsafe modification of image references without the use of the [reference](https://pkg.go.dev/github.com/docker/distribution/reference) package

Reply:
This is driven by KOTS. Their `.airgap` package changes the format of the images, so `docker.io/library/ubuntu:20.04` becomes `<url>/ubuntu:20.04` - note that the "namespace" `library` has been removed.

This is indeed the reason for requiring the `cfg.Repository` value. Because we decided (rightly) to make the use of airgapping implicit in the config, the established pattern has been "if value not equal to the default Gitpod registry".

I looked in the `reference` package and couldn't find a "remove namespace from container URL" function. If I've missed it, please let me know where it is as I'd love to use it

Reply:
Does this affect all images or just the ones from `docker.io/library`?

Reply:
All. I just used that as example

Reply:
That's really bad behaviour on Replicated's part 🤦  We should just get a registry that behaves like a proper mirror.

In that case, rather than making such a blatant exception implicit behaviour, what if we
- tied `RepoName` and `ImageName` to the render context,
- made the "drop repo name" explicit on the render context,
- ask the installer to implement the "drop repo name" behaviour using a command line flag.

Motivation and benefits:
- Makes this odd behaviour explicit: dropping the repository is really unusual. We're only doing this because replicated requires us to. By making this explicit, we'll know in a few months why we're doing this and what triggers it.
- Does not inflate the `ImageName` and `RepoName` API, which means we're not spreading complexity throughout the installer codebase and maintain testability.
- This "drop repo name" behaviour is needed by replicated, in which case we're running the installer from within a container we control. We can pass that flag ourselves. We still keep the config free from this concern.

Reply:
> That's really bad behaviour on Replicated's part facepalm We should just get a registry that behaves like a proper mirror.

Yup

Good idea about dropping the repo name in the config. We're still going to need to change the interface of the `ImageName` and/or `RepoName` functions, but I'll see what I can do 👍🏻

Reply:
I was thinking something akin to:

```Go

func (r *RenderContext) RepoName(repo, name string) string {
	var ref string
	if repo == "" {
		ref = name
	} else {
		ref = fmt.Sprintf("%s/%s", strings.TrimSuffix(repo, "/"), name)
	}
	pref, err := reference.ParseNormalizedNamed(ref)
	if err != nil {
		panic(fmt.Sprintf("cannot parse image repo %s: %v", ref, err))
	}

        if r.DropImageRepo {
		domain := reference.Domain(pref)
                segs := strings.Split(reference.Path(pref), "/")
                return fmt.Sprintf("%s/%s", domain, segs[len(segs) - 1])
        }

	return pref.String()
}

func (r *RenderContext) ImageName(repo, name, tag string) string {
	ref := fmt.Sprintf("%s:%s", r.RepoName(repo, name), tag)
	pref, err := reference.ParseNamed(ref)
	if err != nil {
		panic(fmt.Sprintf("cannot parse image ref %s: %v", ref, err))
	}
	if _, ok := pref.(reference.Tagged); !ok {
		panic(fmt.Sprintf("image ref %s has no tag: %v", ref, err))
	}

	return ref
}
```

Reply:
Changes made and tested